### PR TITLE
Implement the rest of `clean`

### PR DESCRIPTION
**Pull request by TomTaehoonKim**
- [x] Clean invalid branches using CSV
- [x] Implement hard option
- [x] Update comments
- [x] Fix using utils

This PR wraps up the unfinished implementation of `clean` method

### CODE REVIEW

**Comment by TomTaehoonKim** [comment]
I've just seen #148

I need a context update on this issue @YoonhoKim0527 @junha1 (not ASAP)

File: repository/src/lib.rs [hunk 1]
```diff
@@ -188,46 +188,65 @@ impl<T: RawRepository> DistributedRepository<T> {
...
     /// - the `a-#` branches
     /// - the `b-#` branches
-    /// if only the branches are not outdated (branched from the last finalized commit).
-    pub async fn clean(&mut self) -> Result<(), Error> {
+    /// will be left as well
+    /// if only the branches have valid commit sequences
+    /// and are not outdated (branched from the last finalized commit).
+    pub async fn clean(&mut self, hard: bool) -> Result<(), Error> {
         let finalized_branch_commit_hash =
             self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
```
Comment (junha1) [inline repository/src/lib.rs:203]: check integrity error

File: repository/src/lib.rs [hunk 2]
```diff
@@ -188,46 +188,65 @@ impl<T: RawRepository> DistributedRepository<T> {
...
-
-        let branches = self.raw.list_branches().await?;
-
-        // delete outdated p branch, a-# branches, b-# branches
-        for branch in branches {
+        let branches = retrieve_local_branches(&self.raw).await?;
+        let last_header = self.get_last_finalized_block_header().await?;
```
Comment (junha1) [inline repository/src/lib.rs:205]: check integrity error

File: repository/src/lib.rs [hunk 3]
```diff
@@ -188,46 +188,65 @@ impl<T: RawRepository> DistributedRepository<T> {
...
+                        != self
+                            .raw
+                            .find_merge_base(branch_commit_hash, finalized_branch_commit_hash)
+                            .await?
+                    {
+                        self.raw.delete_branch(branch.to_string()).await?;
+                    }
+
+                    // Delete branch with invalid commit sequence
+                    let reserved_state = self.get_reserved_state().await?;
```
Comment (junha1) [inline repository/src/lib.rs:225]: On which commit do you want to get the RS? You must checkout to it first.

File: repository/src/lib.rs (hunk 1 above)
Comment (TomTaehoonKim) [reply repository/src/lib.rs]: Fixed.

File: repository/src/lib.rs (hunk 2 above)
Comment (TomTaehoonKim) [reply repository/src/lib.rs:213]: `get_last_finalized_block_header` already handles integrity error itself. Should I fix this as well?

**Review by YoonhoKim0527 — APPROVED** [review]

File: repository/src/lib.rs (hunk 1 above)
Comment (TomTaehoonKim) [reply repository/src/lib.rs:219]: #326 fixes it.

Comment (TomTaehoonKim) [reply repository/src/lib.rs]: Fixed.

**Pull request merged**